idunn/geocoder/models.py

```diff
@@ -0,0 +1,177 @@
+from enum import Enum
+from typing import Any, List, Optional, Tuple
+
+from pydantic import BaseModel, PositiveInt, confloat
+
+
+Lon = confloat(ge=-180, le=180)
+Lat = confloat(ge=-90, le=90)
+
+Rect = Tuple[Lon, Lat, Lon, Lat]
+
+
+class Coord(BaseModel):
+    lon: Lon
+    lat: Lat
+
+
+class PoiType(BaseModel):
+    id: str
+    name: str
+
+
+class CommercialMode(BaseModel):
+    id: str
+    name: str
+
+
+class PhysicalMode(BaseModel):
+    id: str
+    name: str
+
+
+class Network(BaseModel):
+    id: str
+    name: str
+
+
+class Code(BaseModel):
+    name: str
+    value: str
+
+
+class ZoneType(str, Enum):
+    Suburb = 'suburb'
+    CityDistrict = 'city_district'
+    City = 'city'
+    StateDistrict = 'state_district'
+    State = 'state'
+    CountryRegion = 'country_region'
+    Country = 'country'
+    NonAdministrative = 'non_administrative'
+
+
+class AssociatedAdmin(BaseModel):
+    id: str
+    insee: str
+    level: PositiveInt
+    label: str
+    name: str
+    zip_codes: List[str]
+    coord: Coord
+    bbox: Optional[Rect]
+    zone_type: Optional[ZoneType]
+    parent_id: Optional[str]
+    codes: List[Code]
+
+
+class Line(BaseModel):
+    id: str
+    name: str
+    code: Optional[str]
+    color: Optional[str]
+    text_color: Optional[str]
+    commercial_mode: Optional[CommercialMode]
+    network: Optional[Network]
+    physical_modes: Optional[PhysicalMode]
+    sort_order: Optional[PositiveInt]
+
+
+class FeedPublished(BaseModel):
+    id: str
+    licence: str
+    name: str
+    url: str
+
+
+class GeocodingProperty(BaseModel):
+    key: str
+    value: str
+
+
+class Comment(BaseModel):
+    name: str
+
+
+class GeocodingResponse(BaseModel):
+    type: str
+    label: Optional[str]
+    name: Optional[str]
+    housenumber: Optional[str]
+    street: Optional[str]
+    locality: Optional[str]
+    postcode: Optional[str]
+    city: Optional[str]
+    #  accuracy: Optional[PositiveInt]
+    #  district: Optional[str]
+    #  country: Optional[str]
+    #  state: Optional[str]
+    #  country: Optional[str]
+    #  geohash: Optional[str]
+
+    # From CanalTP/mimirsbrunn:libs/bragi/src/model.rs
+    id: str
+    zone_type: Optional[str]
+    citycode: Optional[str]
+    level: Optional[PositiveInt]
+    administrative_regions: List[AssociatedAdmin]
+    poi_types: List[PoiType] = []
+    properties: List[GeocodingProperty] = []
+    address: Optional['GeocodingResponse']
+    commercial_modes: List[CommercialMode] = []
+    comments: List[Comment] = []
+    physical_modes: List[PhysicalMode] = []
+    lines: List[Line] = []
+    timezone: Optional[str]
+    codes: List[Code] = []
+    feed_publishers: List[FeedPublished] = []
+    bbox: Optional[Rect]
+    country_codes: List[str] = []
+
+GeocodingResponse.update_forward_refs()
+
+
+class FeatureProperties(BaseModel):
+    geocoding: GeocodingResponse
+
+
+class Explaination(BaseModel):
+    value: float
+    description: str
+    details: List['Explaination']
```

Comment:
Isn't it weird to have a type defined as a string?

Reply:
That is kind of weird, but this is the best way I found to fix circular dependancies in the models. 

I just discovered that since Python 3.7 it seems that it is possible to [delay the parsing of type annotations to runtime](https://www.python.org/dev/peps/pep-0563/#resolving-type-hints-at-runtime), however it appears that Idunn should still be 3.6-compatible.